File: app/src/main/java/hu/rayworks/vizit/ui/design/components/VizitDigitalCard.kt
package hu.rayworks.vizit.ui.design.components

import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.aspectRatio
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.material3.Text
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.foundation.layout.height
import androidx.compose.ui.graphics.ImageBitmap
import hu.rayworks.vizit.data.card.CardColorway
import hu.rayworks.vizit.data.card.CardLayout
import hu.rayworks.vizit.data.card.CardPresentation
import hu.rayworks.vizit.ui.design.Vizit

/**
 * The VIZIT digital business card — the product's signature visual.
 *
 * Design intent (Figma: 05 · Card & Profile):
 *  - one object with real presence, not a white rectangle with text on it;
 *  - a navy ink surface with a low-contrast directional gradient that reads as
 *    material, never as decoration;
 *  - a single 4dp cyan edge accent as the brand mark;
 *  - identical in light and dark mode, because the card is the brand.
 */
@Composable
fun VizitDigitalCard(
    fullName: String,
    initials: String,
    jobTitle: String,
    company: String,
    phone: String,
    email: String,
    modifier: Modifier = Modifier,
    photo: ImageBitmap? = null,
    presentation: CardPresentation = CardPresentation(),
    socialLabels: List<String> = emptyList(),
    qrCode: ImageBitmap? = null,
) {
    val colors = Vizit.colors
    val subtitle = listOf(jobTitle, company).filter { it.isNotBlank() }.joinToString(" · ")
    val accent = presentation.colorway.accentColor
    val surface = remember(presentation.colorway) {
        Brush.linearGradient(
            colorStops = arrayOf(
                0.0f to presentation.colorway.startColor,
                0.55f to presentation.colorway.midColor,
                1.0f to presentation.colorway.endColor,
            ),
        )
    }
    val socials = if (presentation.showsSocial) socialLabels else emptyList()

    val describedAs = buildString {
        append(fullName.ifBlank { "Névjegy" })
        if (subtitle.isNotBlank()) append(", ").append(subtitle)
        if (phone.isNotBlank()) append(", telefon ").append(phone)
        if (email.isNotBlank()) append(", e-mail ").append(email)
    }

    Box(
        modifier = modifier
            .fillMaxWidth()
            .aspectRatio(if (presentation.layout == CardLayout.PORTRAIT) 343f / 216f else 343f / 180f)
            .shadow(
                elevation = Vizit.elevation.card,
                shape = RoundedCornerShape(Vizit.radius.xl),
                ambientColor = colors.ink,
                spotColor = colors.ink,
            )
            .clip(RoundedCornerShape(Vizit.radius.xl))
            .background(surface)
            .semantics { contentDescription = describedAs },
    ) {
        // brand edge accent
        Box(
            modifier = Modifier
                .fillMaxHeight()
                .width(4.dp)
                .background(accent),
        )

        val contentModifier = Modifier
            .fillMaxWidth()
            .fillMaxHeight()
            .padding(start = 24.dp, end = 20.dp, top = 22.dp, bottom = 20.dp)

        if (presentation.layout == CardLayout.PORTRAIT) {
            Column(
                modifier = contentModifier,
                verticalArrangement = Arrangement.SpaceBetween,
            ) {
                Row(verticalAlignment = Alignment.CenterVertically) {
                    if (presentation.showsPhoto) {
                        CardAvatar(initials = initials, photo = photo)
                        Spacer(Modifier.width(14.dp))
                    }
                    CardIdentity(fullName = fullName, subtitle = subtitle, modifier = Modifier.weight(1f))
                    if (qrCode != null) {
                        Spacer(Modifier.width(12.dp))
                        CardQr(qrCode)
                    }
                }

                Row(
                    modifier = Modifier.fillMaxWidth(),
                    horizontalArrangement = Arrangement.SpaceBetween,
                    verticalAlignment = Alignment.Bottom,
                ) {
                    CardDetails(phone = phone, email = email, socials = socials)
                    Text(
                        text = "VIZIT",
                        style = Vizit.type.overline.copy(fontWeight = FontWeight.Bold),
                        color = accent,
                    )
                }
            }
        } else {
            Row(modifier = contentModifier) {
                Column(
                    modifier = Modifier.weight(1f).fillMaxHeight(),
                    verticalArrangement = Arrangement.SpaceBetween,
                ) {
                    Row(verticalAlignment = Alignment.CenterVertically) {
                        if (presentation.showsPhoto) {
                            CardAvatar(initials = initials, photo = photo)
                            Spacer(Modifier.width(12.dp))
                        }
                        CardIdentity(fullName = fullName, subtitle = subtitle)
                    }
                    CardDetails(phone = phone, email = email, socials = socials)
                }
                Spacer(Modifier.width(12.dp))
                Column(
                    modifier = Modifier.fillMaxHeight(),
                    horizontalAlignment = Alignment.End,
                    verticalArrangement = Arrangement.SpaceBetween,
                ) {
                    if (qrCode != null) CardQr(qrCode) else Spacer(Modifier.height(1.dp))
                    Text(
                        text = "VIZIT",
                        style = Vizit.type.overline.copy(fontWeight = FontWeight.Bold),
                        color = accent,
                    )
                }
            }
        }
    }
}

@Composable
private fun CardAvatar(initials: String, photo: ImageBitmap?) {
    Box(
        modifier = Modifier
            .size(56.dp)
            .clip(RoundedCornerShape(28.dp))
            .background(Color.White.copy(alpha = 0.12f))
            .border(1.dp, Color.White.copy(alpha = 0.22f), RoundedCornerShape(28.dp)),
        contentAlignment = Alignment.Center,
    ) {
        if (photo != null) {
            Image(
                bitmap = photo,
                contentDescription = null,
                contentScale = ContentScale.Crop,
                modifier = Modifier.fillMaxWidth().clip(RoundedCornerShape(28.dp)),
            )
        } else {
            Text(
                text = initials.ifBlank { "V" },
                style = Vizit.type.title,
                color = Color.White.copy(alpha = 0.92f),
            )
        }
    }
}

@Composable
private fun CardIdentity(fullName: String, subtitle: String, modifier: Modifier = Modifier) {
    Column(modifier = modifier, verticalArrangement = Arrangement.spacedBy(3.dp)) {
        Text(
            text = fullName.ifBlank { "Állítsd össze a névjegyed" },
            style = Vizit.type.h3,
            color = Color.White,
            maxLines = 2,
        )
        if (subtitle.isNotBlank()) {
            Text(
                text = subtitle,
                style = Vizit.type.bodySmall,
                color = Color.White.copy(alpha = 0.68f),
                maxLines = 1,
            )
        }
    }
}

@Composable
private fun CardDetails(phone: String, email: String, socials: List<String>) {
    Column(verticalArrangement = Arrangement.spacedBy(5.dp)) {
        if (phone.isNotBlank()) {
            Text(
                text = phone,
                style = Vizit.type.bodySmall,
                color = Color.White.copy(alpha = 0.82f),
            )
        }
        if (email.isNotBlank()) {
            Text(
                text = email,
                style = Vizit.type.bodySmall,
                color = Color.White.copy(alpha = 0.82f),
                maxLines = 1,
            )
        }
        if (socials.isNotEmpty()) {
            Text(
                text = socials.joinToString(" · "),
                style = Vizit.type.caption,
                color = Color.White.copy(alpha = 0.62f),
                maxLines = 1,
            )
        }
    }
}

/** A miniature of the contact QR, on the white plate a scanner needs. */
@Composable
private fun CardQr(code: ImageBitmap) {
    Box(
        modifier = Modifier
            .size(60.dp)
            .clip(RoundedCornerShape(Vizit.radius.xs))
            .background(Color.White)
            .padding(4.dp),
    ) {
        Image(
            bitmap = code,
            contentDescription = null,
            filterQuality = androidx.compose.ui.graphics.FilterQuality.None,
            modifier = Modifier.fillMaxWidth(),
        )
    }
}

/**
 * The colourway's ARGB values as Compose colours. The presentation model itself
 * stays free of Compose so it can be unit-tested on a plain JVM.
 */
val CardColorway.startColor: Color get() = Color(gradientStart)
val CardColorway.midColor: Color get() = Color(gradientMid)
val CardColorway.endColor: Color get() = Color(gradientEnd)
val CardColorway.accentColor: Color get() = Color(accent)

File: ios/Vizit/Design/VizitDigitalCard.swift
import SwiftUI

/// The VIZIT digital business card — the product's signature visual, identical
/// in intent and metrics to the Android composable of the same name.
///
///  - one object with real presence, not a white rectangle with text on it;
///  - navy ink with a low-contrast directional gradient that reads as material;
///  - a single 4pt cyan edge accent as the brand mark;
///  - the same in light and dark, because the card *is* the brand.
struct VizitDigitalCard: View {
    let profile: ContactProfile
    /// Applied to the name itself, so UI tests can read it as a static text.
    var nameIdentifier: String?
    /// Material, layout and which elements the owner chose in Kártya megjelenése.
    var presentation = CardPresentation()

    private var visible: ContactProfile { profile.visible(through: presentation) }

    private var subtitle: String {
        [visible.jobTitle, visible.company].filter { !$0.isEmpty }.joined(separator: " · ")
    }

    private var accent: Color { Color(uiColor: UIColor(hex: presentation.colorway.accent)) }
    private var primaryText: Color { presentation.colorway.isLight ? VizitColor.ink : .white }
    private var secondaryText: Color {
        presentation.colorway.isLight ? VizitColor.textSecondary : .white.opacity(0.72)
    }

    private var cardAspectRatio: CGFloat {
        switch presentation.layout {
        case .portrait: return 343.0 / 365.0
        case .minimal, .classic: return 343.0 / 216.0
        }
    }

    private var accessibilityText: String {
        var parts = [visible.displayName.isEmpty ? "Névjegy" : visible.displayName]
        if !subtitle.isEmpty { parts.append(subtitle) }
        if !visible.phone.isEmpty { parts.append("telefon \(visible.phone)") }
        if !visible.email.isEmpty { parts.append("e-mail \(visible.email)") }
        return parts.joined(separator: ", ")
    }

    var body: some View {
        // With a name identifier the card is the screen's identity surface, so
        // its name has to stay an addressable element. Without one it reads as
        // a single object, which is the better VoiceOver experience.
        if nameIdentifier == nil {
            surface
                .accessibilityElement(children: .ignore)
                .accessibilityLabel(accessibilityText)
        } else {
            surface.accessibilityElement(children: .contain)
        }
    }

    private var surface: some View {
        ZStack(alignment: .leading) {
            LinearGradient(
                stops: [
                    .init(color: Color(uiColor: UIColor(hex: presentation.colorway.gradient.start)), location: 0),
                    .init(color: Color(uiColor: UIColor(hex: presentation.colorway.gradient.mid)), location: 0.55),
                    .init(color: Color(uiColor: UIColor(hex: presentation.colorway.gradient.end)), location: 1)
                ],
                startPoint: .topLeading,
                endPoint: .bottomTrailing
            )

            Rectangle()
                .fill(accent)
                .frame(width: 4)
                .frame(maxHeight: .infinity)

            if presentation.colorway == .brand {
                Circle()
                    .fill(Color.white.opacity(0.07))
                    .frame(width: 210, height: 210)
                    .offset(x: 225, y: -105)
                    .accessibilityHidden(true)
            }

            Group {
                switch presentation.layout {
                case .portrait:
                    portraitContent
                case .minimal:
                    minimalContent
                case .classic:
                    classicContent
                }
            }
            .padding(.leading, 24)
            .padding(.trailing, 20)
            .padding(.top, 22)
            .padding(.bottom, 20)
        }
        .aspectRatio(cardAspectRatio, contentMode: .fit)
        .clipShape(RoundedRectangle(cornerRadius: VizitRadius.xl, style: .continuous))
        .overlay {
            if presentation.colorway.isLight {
                RoundedRectangle(cornerRadius: VizitRadius.xl, style: .continuous)
                    .stroke(VizitColor.border, lineWidth: 1)
            }
        }
        .vizitShadow(VizitElevation.card)
    }

    /// Portrait is the signature VIZIT composition from the approved screens:
    /// centered identity, calm divider, then the shared contact details.
    private var portraitContent: some View {
        VStack(alignment: .leading, spacing: 0) {
            ZStack {
                avatar(size: 82)
            }
            .frame(maxWidth: .infinity)

            Spacer(minLength: VizitSpace.md)
            portraitIdentity
            Spacer(minLength: VizitSpace.md)

            Rectangle()
                .fill(primaryText.opacity(0.16))
                .frame(height: 1)
                .padding(.horizontal, VizitSpace.md)

            Spacer(minLength: VizitSpace.md)
            portraitDetails
            Spacer(minLength: VizitSpace.md)

            HStack {
                Spacer(minLength: 0)
                Text("VIZIT").vizitOverline().foregroundStyle(accent)
            }
        }
    }

    /// Reduced card face for a calm, contemporary hand-off.
    private var minimalContent: some View {
        VStack(alignment: .leading, spacing: 0) {
            HStack(alignment: .top, spacing: 12) {
                avatar()
                identity
                Spacer(minLength: 0)
            }
            Spacer(minLength: VizitSpace.md)
            HStack(alignment: .bottom) {
                compactDetails
                Spacer(minLength: VizitSpace.sm)
                Text("VIZIT").vizitOverline().foregroundStyle(accent)
            }
        }
    }

    /// Printed-card composition: identity on the leading side and contact
    /// details in a distinct trailing column.
    private var classicContent: some View {
        HStack(alignment: .top, spacing: VizitSpace.md) {
            VStack(alignment: .leading, spacing: VizitSpace.sm) {
                HStack(spacing: 12) {
                    avatar()
                    identity
                }
                Spacer(minLength: 0)
                details
            }
            Spacer(minLength: 0)
            VStack(alignment: .trailing, spacing: VizitSpace.xs) {
                Spacer(minLength: 0)
                Text("VIZIT").vizitOverline().foregroundStyle(accent)
            }
        }
    }

    private var identity: some View {
        VStack(alignment: .leading, spacing: 3) {
            Text(visible.displayName.isEmpty ? "Állítsd össze a névjegyed" : visible.displayName)
                .font(VizitFont.h3)
                .foregroundStyle(primaryText)
                .lineLimit(2)
                .minimumScaleFactor(0.8)
                .vizitIdentifier(nameIdentifier)
            if !subtitle.isEmpty {
                Text(subtitle)
                    .font(VizitFont.bodySmall)
                    .foregroundStyle(secondaryText)
                    .lineLimit(1)
            }
        }
    }

    private var portraitIdentity: some View {
        VStack(spacing: 5) {
            Text(visible.displayName.isEmpty ? "Állítsd össze a névjegyed" : visible.displayName)
                .font(VizitFont.h3)
                .foregroundStyle(primaryText)
                .multilineTextAlignment(.center)
                .lineLimit(2)
                .minimumScaleFactor(0.8)
                .vizitIdentifier(nameIdentifier)
            if !visible.jobTitle.isEmpty {
                Text(visible.jobTitle)
                    .font(VizitFont.caption)
                    .foregroundStyle(accent)
                    .lineLimit(1)
            }
            if !visible.company.isEmpty {
                Text(visible.company)
                    .font(VizitFont.caption)
                    .foregroundStyle(secondaryText)
                    .lineLimit(1)
            }
        }
        .frame(maxWidth: .infinity)
    }

    private var details: some View {
        VStack(alignment: .leading, spacing: 5) {
            if !visible.phone.isEmpty {
                Text(visible.phone)
                    .font(VizitFont.bodySmall)
                    .foregroundStyle(primaryText.opacity(0.84))
            }
            if !visible.email.isEmpty {
                Text(visible.email)
                    .font(VizitFont.bodySmall)
                    .foregroundStyle(primaryText.opacity(0.84))
                    .lineLimit(1)
                    .truncationMode(.middle)
            }
            if !visible.website.isEmpty {
                Text(visible.website)
                    .font(VizitFont.caption)
                    .foregroundStyle(secondaryText)
                    .lineLimit(1)
            }
            if !visible.address.isEmpty {
                Text(visible.address)
                    .font(VizitFont.caption)
                    .foregroundStyle(secondaryText)
                    .lineLimit(1)
            }
        }
    }

    private var compactDetails: some View {
        VStack(alignment: .leading, spacing: 4) {
            if !visible.phone.isEmpty {
                Text(visible.phone)
                    .font(VizitFont.bodySmall)
                    .foregroundStyle(primaryText.opacity(0.84))
            }
            if !visible.email.isEmpty {
                Text(visible.email)
                    .font(VizitFont.caption)
                    .foregroundStyle(secondaryText)
                    .lineLimit(1)
                    .truncationMode(.middle)
            }
        }
    }

    private var portraitDetails: some View {
        VStack(spacing: 5) {
            if !visible.phone.isEmpty {
                Text(visible.phone)
                    .font(VizitFont.bodySmall)
                    .foregroundStyle(primaryText.opacity(0.84))
            }
            if !visible.email.isEmpty {
                Text(visible.email)
                    .font(VizitFont.bodySmall)
                    .foregroundStyle(primaryText.opacity(0.84))
                    .lineLimit(1)
                    .truncationMode(.middle)
            }
            if !visible.website.isEmpty {
                Text(visible.website)
                    .font(VizitFont.caption)
                    .foregroundStyle(secondaryText)
                    .lineLimit(1)
            }
            if !visible.address.isEmpty {
                Text(visible.address)
                    .font(VizitFont.caption)
                    .foregroundStyle(secondaryText)
                    .multilineTextAlignment(.center)
                    .lineLimit(2)
            }
        }
        .frame(maxWidth: .infinity)
    }

    private func avatar(size: CGFloat = 56) -> some View {
        ZStack {
            Circle().fill(primaryText.opacity(presentation.colorway.isLight ? 0.06 : 0.12))
            Circle().stroke(primaryText.opacity(presentation.colorway.isLight ? 0.18 : 0.22), lineWidth: 1)
            if let data = Data(base64Encoded: visible.photoBase64), let image = UIImage(data: data) {
                Image(uiImage: image)
                    .resizable()
                    .scaledToFill()
                    .clipShape(Circle())
            } else {
                Text(visible.initials.isEmpty ? "V" : visible.initials)
                    .font(VizitFont.title)
                    .foregroundStyle(primaryText.opacity(0.92))
            }
        }
        .frame(width: size, height: size)
    }
}

/// Profile photo outside the card, on a themed surface rather than the card's
/// white-on-ink treatment. Falls back to the initials monogram.
struct VizitAvatar: View {
    let profile: ContactProfile
    var size: CGFloat = 72

    var body: some View {
        ZStack {
            Circle().fill(VizitColor.primarySubtle)
            Circle().stroke(VizitColor.border, lineWidth: 1)
            if let data = Data(base64Encoded: profile.photoBase64), let image = UIImage(data: data) {
                Image(uiImage: image)
                    .resizable()
                    .scaledToFill()
                    .clipShape(Circle())
            } else {
                Text(profile.initials.isEmpty ? "V" : profile.initials)
                    .font(.system(size: size * 0.34, weight: .semibold))
                    .foregroundStyle(VizitColor.primary)
            }
        }
        .frame(width: size, height: size)
        .accessibilityHidden(true)
    }
}

/// The official VIZIT artwork always sits on a white plate, in both themes —
/// the logo is a fixed brand asset and must not be tinted or inverted.
struct VizitBrandLockup: View {
    var maxHeight: CGFloat = 132

    private static let image: UIImage? = {
        guard let url = Bundle.main.url(forResource: "VizitLogo", withExtension: "png") else { return nil }
        return UIImage(contentsOfFile: url.path)
    }()

    var body: some View {
        Group {
            if let image = Self.image {
                Image(uiImage: image)
                    .resizable()
                    .scaledToFit()
                    .frame(maxHeight: maxHeight)
                    .padding(VizitSpace.lg)
            } else {
                VStack(spacing: 4) {
                    Text("VIZIT")
                        .font(.system(size: 32, weight: .black))
                        .tracking(7)
                        .foregroundStyle(VizitColor.ink)
                    Text("EGY ÉRINTÉS. EGY KAPCSOLAT.")
                        .vizitOverline()
                        .foregroundStyle(Color(uiColor: UIColor(hex: 0x0B5CE8)))
                }
                .padding(VizitSpace.lg)
            }
        }
        .frame(maxWidth: .infinity)
        .background(Color.white)
        .clipShape(RoundedRectangle(cornerRadius: VizitRadius.xl, style: .continuous))
        .overlay {
            RoundedRectangle(cornerRadius: VizitRadius.xl, style: .continuous)
                .stroke(VizitColor.border, lineWidth: 1)
        }
        .accessibilityElement(children: .ignore)
        .accessibilityLabel("VIZIT – Egy érintés. Egy kapcsolat.")
    }
}
